components/places/src/bookmark_sync/store.rs

```diff
@@ -666,22 +665,36 @@ impl dogear::Driver for Driver {
 }
 
 // The "merger", which is just a thin wrapper for dogear.
-struct Merger<'a> {
+pub(crate) struct Merger<'a> {
     store: &'a BookmarksStore<'a>,
     remote_time: ServerTimestamp,
     local_time: Timestamp,
+    // Used for where the merger is not the one which should be managing the
+    // transaction, e.g. in the case of bookmarks import. The only impact this has
+    // is on the `apply()` function. Always false unless the caller explicitly
+    // turns it on, to avoid accidentally enabling unintentionally.
+    external_transaction: bool,
 }
 
 impl<'a> Merger<'a> {
-    fn new(store: &'a BookmarksStore<'_>, remote_time: ServerTimestamp) -> Self {
+    pub(crate) fn new(store: &'a BookmarksStore<'_>, remote_time: ServerTimestamp) -> Self {
         Self {
             store,
             remote_time,
             local_time: Timestamp::now(),
+            external_transaction: false,
         }
     }
 
-    fn merge(&mut self) -> Result<()> {
+    /// Prevent (or re-enable, in principal) using `begin_transaction` in `apply()`.
+    ///
+    /// The assumption is that if you call this, someone higher up the call_stack is
+    /// managing the transaction at that point.
```

Comment:
What do you think about making this a constructor, like `with_external_transaction`?

Reply:
I think it's niche enough that this use case is fine -- I wanted to avoid it being unclear which thing you should use (all cases other than this should leave it alone). If we want to improve things, maybe the right thing to do would be to let begin_transaction do nothing if a tx is already open.